### ISSUE TITLE
🔄 [Enhancement] 提升代理模型（Surrogate Model）预测准确率

### ISSUE THREAD

**Issue by newtontech** [newtontech/genetic-ml-evolution#30]
## 📋 问题描述

当前 `SurrogateModel` 使用基于 sklearn 的集成方法（Random Forest、Gradient Boosting、MLP）来预测架构性能，但存在以下准确率瓶颈：

### 1. 特征工程不足
- **手工特征编码**：当前使用简单的归一化 + one-hot 编码，仅捕捉浅层结构信息（`surrogate_model.py:80-170`）
- **特征维度不统一**：Transformer (11维)、CNN (14维)、Multimodal (12维)，不同架构类型无法交叉学习
- **缺乏高阶特征交互**：未捕获架构参数之间的组合效应（如层数×隐藏维度的交互）
- **无领域知识注入**：缺少已知的架构设计先验（如参数量、FLOPs、注意力头数与隐藏维度的倍数关系）

### 2. 模型能力有限
- **仅使用传统 ML 模型**：RF/GB/MLP 对非线性高维关系建模能力有限
- **集成策略过于简单**：当前只选最佳单模型，而非真正的加权集成
- **冷启动问题严重**：最少需要 5 个样本才能训练（`surrogate_model.py:196`），早期预测完全随机

### 3. 代理模型使用方式粗糙
- **被动训练**：代理模型仅在 `evolve()` 外部手动训练，未在进化过程中增量更新
- **无不确定性估计**：预测无置信度，无法指导探索-利用平衡
- **未利用进化历史**：每代的中间结果未被充分利用

---

### 方案 A：引入贝叶斯优化（高斯过程回归）
- 使用 **Gaussian Process Regression** 替代/补充当前模型
- 天然提供 **不确定性估计**（预测均值 + 方差）
- 支持主动学习：优先评估不确定性最高的架构
- 适合小样本场景，完美解决冷启动问题

**实现要点：**
```python
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import Matern, WhiteKernel

kernel = Matern(nu=2.5) + WhiteKernel(noise_level=1.0)
gp = GaussianProcessRegressor(kernel=kernel, n_restarts_optimizer=10)
```

### 方案 B：增强特征工程
- **自动特征生成**：添加派生特征（参数量估算、FLOPs 估算、深度×宽度交互项）
- **统一特征空间**：为所有架构类型生成统一长度的特征向量（补零 + 类型标记）
- **目标编码/嵌入**：对离散超参数使用学习到的嵌入而非 one-hot
- **特征选择**：基于互信息或 SHAP 值自动筛选最有信息量的特征

**派生特征示例：**
```python
# Transformer
params_estimate = num_layers * (12 * hidden_size^2 + 12 * hidden_size * ffn_dim)
depth_width_ratio = num_layers * hidden_size / max_seq_len
head_dim = hidden_size / num_heads  # 常见约束：head_dim = 64
```

### 方案 C：增量在线学习 + 加权集成
- **增量训练**：每代进化后自动用新数据更新代理模型（`warm_start=True` 或 partial_fit）
- **加权集成**：基于近期交叉验证分数加权组合多个模型，而非选最优单模型
- **历史数据管理**：引入经验回放缓冲区，保留高质量训练样本

**在 `EvolutionEngine._evaluate_population()` 中集成：**
```python
# 每代结束后增量更新代理模型
if self.surrogate and len(new_evaluations) > 0:
    for arch, fitness in new_evaluations:
        self.surrogate.add_training_point(arch, fitness)
    self.surrogate.fit()  # 增量更新
```

---

### Phase 1: 特征增强（优先级高）
- [ ] 在 `SurrogateModel` 中添加 `_derive_features()` 方法，计算参数量、FLOPs 等派生特征
- [ ] 统一不同架构类型的特征向量长度
- [ ] 添加特征选择逻辑

### Phase 2: 贝叶斯优化集成
- [ ] 添加 `GaussianProcessSurrogate` 类
- [ ] 实现不确定性估计接口 `predict_with_uncertainty()`
- [ ] 在 `EvolutionEngine` 中添加基于不确定性的采样策略（UCB / EI）

### Phase 3: 增量学习
- [ ] 修改 `SurrogateModel.fit()` 支持增量训练
- [ ] 在 `EvolutionEngine.evolve()` 中每代自动更新代理模型
- [ ] 添加经验回放缓冲区管理

### Phase 4: 加权集成
- [ ] 修改 ensemble 策略为加权平均
- [ ] 添加基于滑动窗口的模型权重更新
- [ ] 添加代理模型准确率监控指标

---

## 📊 预期效果

| 指标 | 当前 | 目标 |
|------|------|------|
| 预测 MSE | 基线（未量化） | 降低 30-50% |
| 冷启动样本需求 | 5 个 | 2-3 个（GP） |
| 特征维度 | 11-14（不统一） | 20-30（统一+派生） |
| 不确定性估计 | ❌ 无 | ✅ 有（GP） |
| 在线更新 | ❌ 手动 | ✅ 每代自动 |
| 集成策略 | 选最优单模型 | 加权集成 |

---

## 🏷️ 标签
`enhancement` `surrogate-model` `accuracy` `nas`
